helix-view/src/document.rs

```diff
@@ -457,6 +459,101 @@ where
     *mut_ref = f(mem::take(mut_ref));
 }
 
+#[derive(Clone, Copy, Debug, Default)]
+struct DocumentConfig<T> {
+    pub config: T,
+    pub encoding: Option<&'static encoding::Encoding>,
+    pub line_ending: Option<LineEnding>,
+}
+
+impl<T> DocumentConfig<T> {
+    pub(self) fn transpose(option: Option<DocumentConfig<T>>) -> DocumentConfig<Option<T>> {
+        if let Some(doc_config) = option {
+            DocumentConfig {
+                config: Some(doc_config.config),
+                encoding: doc_config.encoding,
+                line_ending: doc_config.line_ending,
+            }
+        } else {
+            DocumentConfig::default()
+        }
+    }
+}
+
+/// Trait representing methods of configuring the document as it's being opened.
+trait ConfigureDocument {
+    type Config;
+    /// Loads document configuration for a file at `path`.
+    fn load(&self, path: &Path) -> Result<DocumentConfig<Self::Config>, Error>;
+    /// Applies document configuration except for the non-`config` fields of `DocumentConfig`.
+    fn configure_document(doc: &mut Document, settings: Self::Config);
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+struct Autodetect;
+#[derive(Clone, Copy, Debug, Default)]
+struct EditorConfig;
```

Comment:
Can you please leave a comment for these structs? Like `EditorConfig` prioritize editorconfig over user config first.